{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pythonでデータ解析\n",
    "　ここでは、遺伝子発現量ファイルをPythonプログラムで扱う練習をします。ファイルの読み込み、表示、条件にあうデータ行の抽出をおこないます。また、データのグラフ化、統計処理もおこなってみます。  \n",
    "　サンプルデータは、赤シソと青シソの遺伝子発現量のデータです。  \n",
    "\n",
    "---\n",
    "\n",
    "## 0. Jupyter Notebookの基本\n",
    "- 現在のセルのプログラムを実行する: __Ctrl + Enter__    \n",
    "\n",
    "- 現在のセルのプログラムを実行し、次のセルに移る: __Shift + Enter__  \n",
    "- 直前の動作をやり直す: __Ctrl + Z__ (Macの場合 __command + Z__)    \n",
    "誤ってプログラムを消してしまった場合によく使います。  \n",
    "- その他のショートカットを知りたい場合:  __Help > Keyboard Shortcuts__   \n",
    "- ファイルの名前を変更する: __File > Rename__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# 必要なライブラリをインポートする\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. テキストファイル読み込み・表示\n",
    "[読み込むファイル]  \n",
    "* ./05_cuffdiff_out/gene_exp.diff\n",
    "\n",
    "[使用する関数] \n",
    "* ファイルを開く _open()_  \n",
    "* ファイルを閉じる _close()_ \n",
    "* 一行読み込む _readline()_  \n",
    "* 全ての行を読み込む _readlines()_  \n",
    "* 表示する _print()_  \n",
    "* 繰り返し処理をする _for ~ in ...:_  \n",
    "* 条件分岐させる　_if ~: ... else: ..._ \n",
    "* 改行コードを除去する _rstrip()_   \n",
    "* 行をタブコードで分割する: _split('\\t')_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[例]  \n",
    "\n",
    "```python\n",
    "f = open('./05_cuffdiff_out/gene_exp.diff', 'r')\n",
    "lines = f.readlines()\n",
    "f.close()\n",
    "\n",
    "for line in lines:\n",
    "    line = line.rstrip()\n",
    "    items = line.split('\\t')\n",
    "    if 'yes' == items[-1]:\n",
    "        print('This is \"yes\" line')\n",
    "    else:\n",
    "        print('This is \"no\" line')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "---\n",
    "# 2. データフレームを使う\n",
    "[データブレーム]  \n",
    "- Excelのような行列表  \n",
    "\n",
    "[使用する関数]　　\n",
    "- データフレームとしてファイルを読み込む __pd.read_csv(FILE_PATH, sep=..., header=...)__  \n",
    "- 値の数をカウントする __value_counts()__  \n",
    "- 条件にマッチするデータのみ表示する __isin()__\n",
    "- 指定した列のデータを表示する __df.loc__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[例]\n",
    "\n",
    "```python\n",
    "# データフレームとしてファイルを読み込む\n",
    "diff = './05_cuffdiff_out/gene_exp.diff'\n",
    "df = pd.read_csv(diff, sep='\\t', header=0)\n",
    "df\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "```python\n",
    "# 指定した行のデータを表示する\n",
    "df[1:3]\n",
    "\n",
    "# 指定した列のデータを表示する\n",
    "df.loc[:,['locus', 'sample_1', 'sample_2', 'value_1','value_2', 'q_value', 'significant']]\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "```python\n",
    "# 値の数をカウントする\n",
    "df['significant'].value_counts()\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "```python\n",
    "# 条件にマッチするデータのみ表示する\n",
    "sig_yes = df[df['significant'].isin(['yes'])]\n",
    "sig_yes\n",
    "\n",
    "#sig_no = df[df['significant'].isin(['no'])]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. グラフ表示\n",
    "__plt.figure()__, __add_subplot__ など"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# 各座標の準備\n",
    "x = df['value_1']\n",
    "y = df['value_2']\n",
    "\n",
    "# 図の準備\n",
    "#fig = plt.figure()\n",
    "#ax = fig.add_subplot(1,1,1)\n",
    "\n",
    "# 座標データをプロットする\n",
    "#ax.scatter(x,y, c='silver')\n",
    "\n",
    "# 図の装飾（タイトルや軸ラベルなど）\n",
    "#ax.set_title('Gene Expression Plot')\n",
    "#ax.set_xlabel('Gene Expression (FPKM) of red perilla')\n",
    "#ax.set_ylabel('Gene Expression (FPKM) of green perilla')\n",
    "#plt.style.use('ggplot')\n",
    "\n",
    "# 図の描画\n",
    "#fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[例1] 散布図\n",
    "\n",
    "```\n",
    "# 各座標の準備\n",
    "# \"yes\"データの(x座標, y座標)=(value_1, value_2)を取得する。\n",
    "x1 = sig_yes['value_1']\n",
    "y1 = sig_yes['value_2']\n",
    "# \"no\"データの(x座標, y座標)=(value_1, value_2)を取得する。\n",
    "x2 = sig_no['value_1']\n",
    "y2 = sig_no['value_2']\n",
    "\n",
    "# 対数値に変換 (log)\n",
    "x1 = np.log10(x1 + 10**(-10))\n",
    "y1 = np.log10(y1 + 10**(-10))\n",
    "x2 = np.log10(x2 + 10**(-10))\n",
    "y2 = np.log10(y2 + 10**(-10))\n",
    "\n",
    "# 図の準備\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(1,1,1)\n",
    "\n",
    "# 座標データをプロットする\n",
    "ax.scatter(x2,y2, c='silver')\n",
    "ax.scatter(x1,y1, c='red')\n",
    "\n",
    "# 図の装飾（タイトルや軸ラベルなど）\n",
    "ax.set_title('Gene Expression Plot')\n",
    "ax.set_xlabel('Gene Expression (FPKM) of red perilla')\n",
    "ax.set_ylabel('Gene Expression (FPKM) of green perilla')\n",
    "plt.style.use('ggplot')\n",
    "\n",
    "# 図の描画\n",
    "fig.show()\n",
    "```\n",
    "---\n",
    "![fpkm_scatterplot](./etc/fpkm_scatterplot.png \"FPKM scatter plot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[例2] 箱ひげ図\n",
    "\n",
    "```python\n",
    "# データ準備\n",
    "red = list(np.log10(df['value_1'] + 10**(-10)))\n",
    "green = list(np.log10(df['value_2'] + 10**(-10)))\n",
    "\n",
    "# 図の準備\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(1,1,1)\n",
    "\n",
    "# 箱ひげ図にデータをプロットする\n",
    "cmp = (red, green)\n",
    "bp = ax.boxplot(cmp)\n",
    "\n",
    "# 図の装飾\n",
    "plt.title('Box plot of Gene Expression')\n",
    "plt.ylabel('log10 of FPKM')\n",
    "ax.set_xticklabels(['red perilla', 'green perilla'])\n",
    "plt.style.use('ggplot')\n",
    "\n",
    "# 図の描画\n",
    "plt.show()\n",
    "```\n",
    "---\n",
    "![fpkm_boxplot](./etc/fpkm_boxplot.png \"FPKM box plot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. 統計処理\n",
    "対応のある２群のt検定 __stats.ttest_rel()__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c10478_g1_i1:4-878</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>31265.30</td>\n",
       "      <td>39514.70</td>\n",
       "      <td>0.096678</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c1235_g1_i1:0-358</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>18899.00</td>\n",
       "      <td>25190.00</td>\n",
       "      <td>0.450315</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c14532_g1_i1:672-1250</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>1036.71</td>\n",
       "      <td>6905.89</td>\n",
       "      <td>0.098630</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c14964_g1_i1:0-1591</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>96498.60</td>\n",
       "      <td>106401.00</td>\n",
       "      <td>0.096678</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c18250_g2_i1:33-1490</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>139165.00</td>\n",
       "      <td>27992.50</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>c18250_g2_i2:33-1519</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>135968.00</td>\n",
       "      <td>24295.30</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>c20303_g1_i1:0-1582</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>255968.00</td>\n",
       "      <td>55893.00</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c2030_g1_i1:0-433</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>21027.10</td>\n",
       "      <td>10418.10</td>\n",
       "      <td>0.096678</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>c21075_g2_i1:0-1482</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>485920.00</td>\n",
       "      <td>107958.00</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>c22175_g1_i1:0-492</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>16423.80</td>\n",
       "      <td>29570.80</td>\n",
       "      <td>0.076160</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>c26519_g1_i1:0-387</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>24600.00</td>\n",
       "      <td>27639.20</td>\n",
       "      <td>0.686500</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>c33987_g1_i1:0-287</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>55510.20</td>\n",
       "      <td>25875.90</td>\n",
       "      <td>0.096678</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>c41275_g1_i1:0-688</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>47475.00</td>\n",
       "      <td>41485.00</td>\n",
       "      <td>0.346383</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>c8983_g1_i1:0-396</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>5659.34</td>\n",
       "      <td>18778.50</td>\n",
       "      <td>0.100609</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    locus     sample_1       sample_2    value_1    value_2  \\\n",
       "0      c10478_g1_i1:4-878  red_perilla  green_perilla   31265.30   39514.70   \n",
       "1       c1235_g1_i1:0-358  red_perilla  green_perilla   18899.00   25190.00   \n",
       "2   c14532_g1_i1:672-1250  red_perilla  green_perilla    1036.71    6905.89   \n",
       "3     c14964_g1_i1:0-1591  red_perilla  green_perilla   96498.60  106401.00   \n",
       "4    c18250_g2_i1:33-1490  red_perilla  green_perilla  139165.00   27992.50   \n",
       "5    c18250_g2_i2:33-1519  red_perilla  green_perilla  135968.00   24295.30   \n",
       "6     c20303_g1_i1:0-1582  red_perilla  green_perilla  255968.00   55893.00   \n",
       "7       c2030_g1_i1:0-433  red_perilla  green_perilla   21027.10   10418.10   \n",
       "8     c21075_g2_i1:0-1482  red_perilla  green_perilla  485920.00  107958.00   \n",
       "9      c22175_g1_i1:0-492  red_perilla  green_perilla   16423.80   29570.80   \n",
       "10     c26519_g1_i1:0-387  red_perilla  green_perilla   24600.00   27639.20   \n",
       "11     c33987_g1_i1:0-287  red_perilla  green_perilla   55510.20   25875.90   \n",
       "12     c41275_g1_i1:0-688  red_perilla  green_perilla   47475.00   41485.00   \n",
       "13      c8983_g1_i1:0-396  red_perilla  green_perilla    5659.34   18778.50   \n",
       "\n",
       "     q_value significant  \n",
       "0   0.096678          no  \n",
       "1   0.450315          no  \n",
       "2   0.098630          no  \n",
       "3   0.096678          no  \n",
       "4   0.000175         yes  \n",
       "5   0.000175         yes  \n",
       "6   0.000175         yes  \n",
       "7   0.096678          no  \n",
       "8   0.000175         yes  \n",
       "9   0.076160          no  \n",
       "10  0.686500          no  \n",
       "11  0.096678          no  \n",
       "12  0.346383          no  \n",
       "13  0.100609          no  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[:,['locus', 'sample_1', 'sample_2', 'value_1','value_2', 'q_value', 'significant']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=1.8679404345702795, pvalue=0.084472971958650486)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "red = list(df['value_1'])\n",
    "green = list(df['value_2'])\n",
    "res = stats.ttest_rel(red, green)\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 5. 正規表現で文字列の置換\n",
    "正規表現で文字列置換をする __re.sub()__  \n",
    "リストに追加する __append()__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "      <th>tr_locus</th>\n",
       "      <th>new_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>-</td>\n",
       "      <td>c10478_g1_i1:4-878</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>28181.7</td>\n",
       "      <td>51973.2</td>\n",
       "      <td>0.883007</td>\n",
       "      <td>4.60533</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c10478_g1_i1</td>\n",
       "      <td>c10478_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>c1235_g1_i1:1-358</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>17157.9</td>\n",
       "      <td>31271.0</td>\n",
       "      <td>0.865952</td>\n",
       "      <td>1.57945</td>\n",
       "      <td>0.11425</td>\n",
       "      <td>0.152333</td>\n",
       "      <td>no</td>\n",
       "      <td>c1235_g1_i1</td>\n",
       "      <td>c1235_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>-</td>\n",
       "      <td>c14964_g1_i1:1-1591</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>87044.0</td>\n",
       "      <td>139830.0</td>\n",
       "      <td>0.683862</td>\n",
       "      <td>8.79177</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c14964_g1_i1</td>\n",
       "      <td>c14964_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>-</td>\n",
       "      <td>c18250_g2_i1:33-1490</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>125438.0</td>\n",
       "      <td>36818.1</td>\n",
       "      <td>-1.768490</td>\n",
       "      <td>-15.77070</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c18250_g2_i1</td>\n",
       "      <td>c18250_g2_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>-</td>\n",
       "      <td>c18250_g2_i2:33-1519</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>122557.0</td>\n",
       "      <td>31955.3</td>\n",
       "      <td>-1.939320</td>\n",
       "      <td>-16.49500</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c18250_g2_i2</td>\n",
       "      <td>c18250_g2_i2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>XLOC_000006</td>\n",
       "      <td>XLOC_000006</td>\n",
       "      <td>-</td>\n",
       "      <td>c20303_g1_i1:0-1582</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>230720.0</td>\n",
       "      <td>73515.3</td>\n",
       "      <td>-1.650030</td>\n",
       "      <td>-21.72600</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c20303_g1_i1</td>\n",
       "      <td>c20303_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>XLOC_000007</td>\n",
       "      <td>XLOC_000007</td>\n",
       "      <td>-</td>\n",
       "      <td>c2030_g1_i1:0-423</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>19894.8</td>\n",
       "      <td>11484.1</td>\n",
       "      <td>-0.792752</td>\n",
       "      <td>-1.44577</td>\n",
       "      <td>0.16005</td>\n",
       "      <td>0.174600</td>\n",
       "      <td>no</td>\n",
       "      <td>c2030_g1_i1</td>\n",
       "      <td>c2030_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>-</td>\n",
       "      <td>c21075_g2_i1:0-1482</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>437991.0</td>\n",
       "      <td>141996.0</td>\n",
       "      <td>-1.625050</td>\n",
       "      <td>-28.55720</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c21075_g2_i1</td>\n",
       "      <td>c21075_g2_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>XLOC_000009</td>\n",
       "      <td>XLOC_000009</td>\n",
       "      <td>-</td>\n",
       "      <td>c22175_g1_i1:0-492</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>14804.5</td>\n",
       "      <td>38894.0</td>\n",
       "      <td>1.393510</td>\n",
       "      <td>3.53702</td>\n",
       "      <td>0.00240</td>\n",
       "      <td>0.004114</td>\n",
       "      <td>yes</td>\n",
       "      <td>c22175_g1_i1</td>\n",
       "      <td>c22175_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>XLOC_000010</td>\n",
       "      <td>XLOC_000010</td>\n",
       "      <td>-</td>\n",
       "      <td>c26519_g1_i1:3-387</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>22579.1</td>\n",
       "      <td>35223.8</td>\n",
       "      <td>0.641562</td>\n",
       "      <td>1.45503</td>\n",
       "      <td>0.14975</td>\n",
       "      <td>0.174600</td>\n",
       "      <td>no</td>\n",
       "      <td>c26519_g1_i1</td>\n",
       "      <td>c26519_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>XLOC_000011</td>\n",
       "      <td>XLOC_000011</td>\n",
       "      <td>-</td>\n",
       "      <td>c33987_g1_i1:0-287</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>50042.3</td>\n",
       "      <td>34034.2</td>\n",
       "      <td>-0.556161</td>\n",
       "      <td>-1.02444</td>\n",
       "      <td>0.29920</td>\n",
       "      <td>0.299200</td>\n",
       "      <td>no</td>\n",
       "      <td>c33987_g1_i1</td>\n",
       "      <td>c33987_g1_i1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>-</td>\n",
       "      <td>c41275_g1_i1:0-688</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>42793.2</td>\n",
       "      <td>54564.7</td>\n",
       "      <td>0.350588</td>\n",
       "      <td>1.78003</td>\n",
       "      <td>0.07705</td>\n",
       "      <td>0.115575</td>\n",
       "      <td>no</td>\n",
       "      <td>c41275_g1_i1</td>\n",
       "      <td>c41275_g1_i1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id gene                 locus     sample_1  \\\n",
       "0   XLOC_000001  XLOC_000001    -    c10478_g1_i1:4-878  red_perilla   \n",
       "1   XLOC_000002  XLOC_000002    -     c1235_g1_i1:1-358  red_perilla   \n",
       "2   XLOC_000003  XLOC_000003    -   c14964_g1_i1:1-1591  red_perilla   \n",
       "3   XLOC_000004  XLOC_000004    -  c18250_g2_i1:33-1490  red_perilla   \n",
       "4   XLOC_000005  XLOC_000005    -  c18250_g2_i2:33-1519  red_perilla   \n",
       "5   XLOC_000006  XLOC_000006    -   c20303_g1_i1:0-1582  red_perilla   \n",
       "6   XLOC_000007  XLOC_000007    -     c2030_g1_i1:0-423  red_perilla   \n",
       "7   XLOC_000008  XLOC_000008    -   c21075_g2_i1:0-1482  red_perilla   \n",
       "8   XLOC_000009  XLOC_000009    -    c22175_g1_i1:0-492  red_perilla   \n",
       "9   XLOC_000010  XLOC_000010    -    c26519_g1_i1:3-387  red_perilla   \n",
       "10  XLOC_000011  XLOC_000011    -    c33987_g1_i1:0-287  red_perilla   \n",
       "11  XLOC_000012  XLOC_000012    -    c41275_g1_i1:0-688  red_perilla   \n",
       "\n",
       "         sample_2 status   value_1   value_2  log2(fold_change)  test_stat  \\\n",
       "0   green_perilla     OK   28181.7   51973.2           0.883007    4.60533   \n",
       "1   green_perilla     OK   17157.9   31271.0           0.865952    1.57945   \n",
       "2   green_perilla     OK   87044.0  139830.0           0.683862    8.79177   \n",
       "3   green_perilla     OK  125438.0   36818.1          -1.768490  -15.77070   \n",
       "4   green_perilla     OK  122557.0   31955.3          -1.939320  -16.49500   \n",
       "5   green_perilla     OK  230720.0   73515.3          -1.650030  -21.72600   \n",
       "6   green_perilla     OK   19894.8   11484.1          -0.792752   -1.44577   \n",
       "7   green_perilla     OK  437991.0  141996.0          -1.625050  -28.55720   \n",
       "8   green_perilla     OK   14804.5   38894.0           1.393510    3.53702   \n",
       "9   green_perilla     OK   22579.1   35223.8           0.641562    1.45503   \n",
       "10  green_perilla     OK   50042.3   34034.2          -0.556161   -1.02444   \n",
       "11  green_perilla     OK   42793.2   54564.7           0.350588    1.78003   \n",
       "\n",
       "    p_value   q_value significant      tr_locus     new_locus  \n",
       "0   0.00005  0.000100         yes  c10478_g1_i1  c10478_g1_i1  \n",
       "1   0.11425  0.152333          no   c1235_g1_i1   c1235_g1_i1  \n",
       "2   0.00005  0.000100         yes  c14964_g1_i1  c14964_g1_i1  \n",
       "3   0.00005  0.000100         yes  c18250_g2_i1  c18250_g2_i1  \n",
       "4   0.00005  0.000100         yes  c18250_g2_i2  c18250_g2_i2  \n",
       "5   0.00005  0.000100         yes  c20303_g1_i1  c20303_g1_i1  \n",
       "6   0.16005  0.174600          no   c2030_g1_i1   c2030_g1_i1  \n",
       "7   0.00005  0.000100         yes  c21075_g2_i1  c21075_g2_i1  \n",
       "8   0.00240  0.004114         yes  c22175_g1_i1  c22175_g1_i1  \n",
       "9   0.14975  0.174600          no  c26519_g1_i1  c26519_g1_i1  \n",
       "10  0.29920  0.299200          no  c33987_g1_i1  c33987_g1_i1  \n",
       "11  0.07705  0.115575          no  c41275_g1_i1  c41275_g1_i1  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 置換した文字列を一時的に保管するリスト\n",
    "temp = []\n",
    "\n",
    "# locus列のc+番号の部分のみ抽出 = :(コロン)以降の文字列を除去する\n",
    "for name in df['locus']:\n",
    "    rep_name = re.sub(':[0-9\\-]+$', '', name)\n",
    "    temp.append(rep_name)\n",
    "\n",
    "# データフレームに置換した文字列の列'new_locus'を追加する\n",
    "df['new_locus'] = temp\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. 二つのデータフレームを連結する\n",
    "データフレームを連結する __pd.merge()__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>EC_no.</th>\n",
       "      <th>Enzyme_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c1235_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c2030_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c8983_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c10478_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c11655_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>c11655_g1_i2</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>c14532_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c14532_g1_i2</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>c14964_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>c18250_g2_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>c18250_g2_i2</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>c20303_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>c21075_g2_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>c22175_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>c22598_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>c24207_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>c24350_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>c26519_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c33198_g1_i1</td>\n",
       "      <td>2.3.1.171;2.3.1.172</td>\n",
       "      <td>anthocyanin 6''-O-malonyltransferase;anthocyan...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>c33987_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>c34984_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>c36176_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>c37800_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>c41275_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>c42167_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Contig               EC_no.  \\\n",
       "0    c1235_g1_i1            2.4.1.115   \n",
       "1    c2030_g1_i1            2.3.1.172   \n",
       "2    c8983_g1_i1            2.4.1.115   \n",
       "3   c10478_g1_i1            2.3.1.172   \n",
       "4   c11655_g1_i1            2.3.1.172   \n",
       "5   c11655_g1_i2            2.3.1.172   \n",
       "6   c14532_g1_i1            2.3.1.172   \n",
       "7   c14532_g1_i2            2.3.1.172   \n",
       "8   c14964_g1_i1            2.3.1.172   \n",
       "9   c18250_g2_i1            2.3.1.172   \n",
       "10  c18250_g2_i2            2.3.1.172   \n",
       "11  c20303_g1_i1            2.4.1.115   \n",
       "12  c21075_g2_i1            2.3.1.172   \n",
       "13  c22175_g1_i1            2.3.1.172   \n",
       "14  c22598_g1_i1            2.4.1.115   \n",
       "15  c24207_g1_i1            2.3.1.172   \n",
       "16  c24350_g1_i1            2.3.1.172   \n",
       "17  c26519_g1_i1            2.3.1.172   \n",
       "18  c33198_g1_i1  2.3.1.171;2.3.1.172   \n",
       "19  c33987_g1_i1            2.4.1.115   \n",
       "20  c34984_g1_i1            2.4.1.115   \n",
       "21  c36176_g1_i1            2.3.1.172   \n",
       "22  c37800_g1_i1            2.3.1.172   \n",
       "23  c41275_g1_i1            2.3.1.172   \n",
       "24  c42167_g1_i1            2.3.1.172   \n",
       "\n",
       "                                          Enzyme_name  \n",
       "0               anthocyanidin 3-O-glucosyltransferase  \n",
       "1   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "2               anthocyanidin 3-O-glucosyltransferase  \n",
       "3   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "4   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "5   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "6   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "7   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "8   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "9   anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "10  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "11              anthocyanidin 3-O-glucosyltransferase  \n",
       "12  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "13  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "14              anthocyanidin 3-O-glucosyltransferase  \n",
       "15  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "16  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "17  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "18  anthocyanin 6''-O-malonyltransferase;anthocyan...  \n",
       "19              anthocyanidin 3-O-glucosyltransferase  \n",
       "20              anthocyanidin 3-O-glucosyltransferase  \n",
       "21  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "22  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "23  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "24  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 新しいファイルを読み込み\n",
    "ctg = pd.read_csv(\"./etc/reference_contigs.txt\", sep='\\t', header=0)\n",
    "ctg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "      <th>tr_locus</th>\n",
       "      <th>new_locus</th>\n",
       "      <th>Contig</th>\n",
       "      <th>EC_no.</th>\n",
       "      <th>Enzyme_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>-</td>\n",
       "      <td>c10478_g1_i1:4-878</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>28181.7</td>\n",
       "      <td>51973.2</td>\n",
       "      <td>0.883007</td>\n",
       "      <td>4.60533</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c10478_g1_i1</td>\n",
       "      <td>c10478_g1_i1</td>\n",
       "      <td>c10478_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>c1235_g1_i1:1-358</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>17157.9</td>\n",
       "      <td>31271.0</td>\n",
       "      <td>0.865952</td>\n",
       "      <td>1.57945</td>\n",
       "      <td>0.11425</td>\n",
       "      <td>0.152333</td>\n",
       "      <td>no</td>\n",
       "      <td>c1235_g1_i1</td>\n",
       "      <td>c1235_g1_i1</td>\n",
       "      <td>c1235_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>-</td>\n",
       "      <td>c14964_g1_i1:1-1591</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>87044.0</td>\n",
       "      <td>139830.0</td>\n",
       "      <td>0.683862</td>\n",
       "      <td>8.79177</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c14964_g1_i1</td>\n",
       "      <td>c14964_g1_i1</td>\n",
       "      <td>c14964_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>-</td>\n",
       "      <td>c18250_g2_i1:33-1490</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>125438.0</td>\n",
       "      <td>36818.1</td>\n",
       "      <td>-1.768490</td>\n",
       "      <td>-15.77070</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c18250_g2_i1</td>\n",
       "      <td>c18250_g2_i1</td>\n",
       "      <td>c18250_g2_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>-</td>\n",
       "      <td>c18250_g2_i2:33-1519</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>122557.0</td>\n",
       "      <td>31955.3</td>\n",
       "      <td>-1.939320</td>\n",
       "      <td>-16.49500</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c18250_g2_i2</td>\n",
       "      <td>c18250_g2_i2</td>\n",
       "      <td>c18250_g2_i2</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>XLOC_000006</td>\n",
       "      <td>XLOC_000006</td>\n",
       "      <td>-</td>\n",
       "      <td>c20303_g1_i1:0-1582</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>230720.0</td>\n",
       "      <td>73515.3</td>\n",
       "      <td>-1.650030</td>\n",
       "      <td>-21.72600</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c20303_g1_i1</td>\n",
       "      <td>c20303_g1_i1</td>\n",
       "      <td>c20303_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>XLOC_000007</td>\n",
       "      <td>XLOC_000007</td>\n",
       "      <td>-</td>\n",
       "      <td>c2030_g1_i1:0-423</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>19894.8</td>\n",
       "      <td>11484.1</td>\n",
       "      <td>-0.792752</td>\n",
       "      <td>-1.44577</td>\n",
       "      <td>0.16005</td>\n",
       "      <td>0.174600</td>\n",
       "      <td>no</td>\n",
       "      <td>c2030_g1_i1</td>\n",
       "      <td>c2030_g1_i1</td>\n",
       "      <td>c2030_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>-</td>\n",
       "      <td>c21075_g2_i1:0-1482</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>437991.0</td>\n",
       "      <td>141996.0</td>\n",
       "      <td>-1.625050</td>\n",
       "      <td>-28.55720</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>yes</td>\n",
       "      <td>c21075_g2_i1</td>\n",
       "      <td>c21075_g2_i1</td>\n",
       "      <td>c21075_g2_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>XLOC_000009</td>\n",
       "      <td>XLOC_000009</td>\n",
       "      <td>-</td>\n",
       "      <td>c22175_g1_i1:0-492</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>14804.5</td>\n",
       "      <td>38894.0</td>\n",
       "      <td>1.393510</td>\n",
       "      <td>3.53702</td>\n",
       "      <td>0.00240</td>\n",
       "      <td>0.004114</td>\n",
       "      <td>yes</td>\n",
       "      <td>c22175_g1_i1</td>\n",
       "      <td>c22175_g1_i1</td>\n",
       "      <td>c22175_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>XLOC_000010</td>\n",
       "      <td>XLOC_000010</td>\n",
       "      <td>-</td>\n",
       "      <td>c26519_g1_i1:3-387</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>22579.1</td>\n",
       "      <td>35223.8</td>\n",
       "      <td>0.641562</td>\n",
       "      <td>1.45503</td>\n",
       "      <td>0.14975</td>\n",
       "      <td>0.174600</td>\n",
       "      <td>no</td>\n",
       "      <td>c26519_g1_i1</td>\n",
       "      <td>c26519_g1_i1</td>\n",
       "      <td>c26519_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>XLOC_000011</td>\n",
       "      <td>XLOC_000011</td>\n",
       "      <td>-</td>\n",
       "      <td>c33987_g1_i1:0-287</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>50042.3</td>\n",
       "      <td>34034.2</td>\n",
       "      <td>-0.556161</td>\n",
       "      <td>-1.02444</td>\n",
       "      <td>0.29920</td>\n",
       "      <td>0.299200</td>\n",
       "      <td>no</td>\n",
       "      <td>c33987_g1_i1</td>\n",
       "      <td>c33987_g1_i1</td>\n",
       "      <td>c33987_g1_i1</td>\n",
       "      <td>2.4.1.115</td>\n",
       "      <td>anthocyanidin 3-O-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>-</td>\n",
       "      <td>c41275_g1_i1:0-688</td>\n",
       "      <td>red_perilla</td>\n",
       "      <td>green_perilla</td>\n",
       "      <td>OK</td>\n",
       "      <td>42793.2</td>\n",
       "      <td>54564.7</td>\n",
       "      <td>0.350588</td>\n",
       "      <td>1.78003</td>\n",
       "      <td>0.07705</td>\n",
       "      <td>0.115575</td>\n",
       "      <td>no</td>\n",
       "      <td>c41275_g1_i1</td>\n",
       "      <td>c41275_g1_i1</td>\n",
       "      <td>c41275_g1_i1</td>\n",
       "      <td>2.3.1.172</td>\n",
       "      <td>anthocyanin 5-O-glucoside 6'''-O-malonyltransf...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id gene                 locus     sample_1  \\\n",
       "0   XLOC_000001  XLOC_000001    -    c10478_g1_i1:4-878  red_perilla   \n",
       "1   XLOC_000002  XLOC_000002    -     c1235_g1_i1:1-358  red_perilla   \n",
       "2   XLOC_000003  XLOC_000003    -   c14964_g1_i1:1-1591  red_perilla   \n",
       "3   XLOC_000004  XLOC_000004    -  c18250_g2_i1:33-1490  red_perilla   \n",
       "4   XLOC_000005  XLOC_000005    -  c18250_g2_i2:33-1519  red_perilla   \n",
       "5   XLOC_000006  XLOC_000006    -   c20303_g1_i1:0-1582  red_perilla   \n",
       "6   XLOC_000007  XLOC_000007    -     c2030_g1_i1:0-423  red_perilla   \n",
       "7   XLOC_000008  XLOC_000008    -   c21075_g2_i1:0-1482  red_perilla   \n",
       "8   XLOC_000009  XLOC_000009    -    c22175_g1_i1:0-492  red_perilla   \n",
       "9   XLOC_000010  XLOC_000010    -    c26519_g1_i1:3-387  red_perilla   \n",
       "10  XLOC_000011  XLOC_000011    -    c33987_g1_i1:0-287  red_perilla   \n",
       "11  XLOC_000012  XLOC_000012    -    c41275_g1_i1:0-688  red_perilla   \n",
       "\n",
       "         sample_2 status   value_1   value_2  log2(fold_change)  test_stat  \\\n",
       "0   green_perilla     OK   28181.7   51973.2           0.883007    4.60533   \n",
       "1   green_perilla     OK   17157.9   31271.0           0.865952    1.57945   \n",
       "2   green_perilla     OK   87044.0  139830.0           0.683862    8.79177   \n",
       "3   green_perilla     OK  125438.0   36818.1          -1.768490  -15.77070   \n",
       "4   green_perilla     OK  122557.0   31955.3          -1.939320  -16.49500   \n",
       "5   green_perilla     OK  230720.0   73515.3          -1.650030  -21.72600   \n",
       "6   green_perilla     OK   19894.8   11484.1          -0.792752   -1.44577   \n",
       "7   green_perilla     OK  437991.0  141996.0          -1.625050  -28.55720   \n",
       "8   green_perilla     OK   14804.5   38894.0           1.393510    3.53702   \n",
       "9   green_perilla     OK   22579.1   35223.8           0.641562    1.45503   \n",
       "10  green_perilla     OK   50042.3   34034.2          -0.556161   -1.02444   \n",
       "11  green_perilla     OK   42793.2   54564.7           0.350588    1.78003   \n",
       "\n",
       "    p_value   q_value significant      tr_locus     new_locus        Contig  \\\n",
       "0   0.00005  0.000100         yes  c10478_g1_i1  c10478_g1_i1  c10478_g1_i1   \n",
       "1   0.11425  0.152333          no   c1235_g1_i1   c1235_g1_i1   c1235_g1_i1   \n",
       "2   0.00005  0.000100         yes  c14964_g1_i1  c14964_g1_i1  c14964_g1_i1   \n",
       "3   0.00005  0.000100         yes  c18250_g2_i1  c18250_g2_i1  c18250_g2_i1   \n",
       "4   0.00005  0.000100         yes  c18250_g2_i2  c18250_g2_i2  c18250_g2_i2   \n",
       "5   0.00005  0.000100         yes  c20303_g1_i1  c20303_g1_i1  c20303_g1_i1   \n",
       "6   0.16005  0.174600          no   c2030_g1_i1   c2030_g1_i1   c2030_g1_i1   \n",
       "7   0.00005  0.000100         yes  c21075_g2_i1  c21075_g2_i1  c21075_g2_i1   \n",
       "8   0.00240  0.004114         yes  c22175_g1_i1  c22175_g1_i1  c22175_g1_i1   \n",
       "9   0.14975  0.174600          no  c26519_g1_i1  c26519_g1_i1  c26519_g1_i1   \n",
       "10  0.29920  0.299200          no  c33987_g1_i1  c33987_g1_i1  c33987_g1_i1   \n",
       "11  0.07705  0.115575          no  c41275_g1_i1  c41275_g1_i1  c41275_g1_i1   \n",
       "\n",
       "       EC_no.                                        Enzyme_name  \n",
       "0   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "1   2.4.1.115              anthocyanidin 3-O-glucosyltransferase  \n",
       "2   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "3   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "4   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "5   2.4.1.115              anthocyanidin 3-O-glucosyltransferase  \n",
       "6   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "7   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "8   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "9   2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  \n",
       "10  2.4.1.115              anthocyanidin 3-O-glucosyltransferase  \n",
       "11  2.3.1.172  anthocyanin 5-O-glucoside 6'''-O-malonyltransf...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 遺伝子発現のデータフレームの'new_locus'の文字列と新しいファイルの'Contig'の文字列をキーに、連結する。\n",
    "df_plus = pd.merge(df, ctg, left_on='new_locus', right_on='Contig', how='left')\n",
    "df_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
